app/src/main/java/org/mozilla/rocket/download/DownloadInfoViewModel.kt

```diff
@@ -0,0 +1,264 @@
+                val file = File(URI.create(download.fileUri).path)
+                if (file.exists()) {
+                    deleteSnackbar.value = download
+                } else {
+                    toastMessage.value = R.string.cannot_find_the_file
+                }
+            }
+        })
+    }
+
+    fun confirmDelete(download: DownloadInfo) {
+        val deleteFile = File(URI.create(download.fileUri).path)
+        try {
+            if (deleteFile.delete()) {
+                repository.deleteFromDownloadManager(download.downloadId)
+                repository.remove(download.rowId)
+            } else {
+                toastMessage.value = R.string.cannot_delete_the_file
+            }
+        } catch (e: Exception) {
+            Log.e(this.javaClass.simpleName, "" + e.message)
+            toastMessage.value = R.string.cannot_delete_the_file
+        }
+    }
+
+    fun add(downloadInfo: DownloadInfo) {
+        var index = -1
+        for (i in 0 until downloadInfoBundle.list.size) {
+            if (downloadInfoBundle.list[i].rowId < downloadInfo.rowId) {
+                index = i
+                break
+            }
+        }
+
+        if (index == -1) {
+            downloadInfoBundle.list.add(downloadInfo)
+            //The crash will happen when data set size is 1 after add item.
+            //Because we define item count is 1 and mDownloadInfo is empty that means nothing and show empty view.
+            //So use notifyDataSetChanged() instead of notifyItemInserted when data size is 1 after add item.
+            if (downloadInfoBundle.list.size > 1) {
+                downloadInfoBundle.notifyType = DownloadInfoBundle.Constants.NOTIFY_ITEM_INSERTED
+                downloadInfoBundle.index = (downloadInfoBundle.list.size - 1).toLong()
+            } else {
+                downloadInfoBundle.notifyType = DownloadInfoBundle.Constants.NOTIFY_DATASET_CHANGED
+            }
+        } else {
+            downloadInfoBundle.list.add(index, downloadInfo)
+            downloadInfoBundle.notifyType = DownloadInfoBundle.Constants.NOTIFY_ITEM_INSERTED
+            downloadInfoBundle.index = index.toLong()
+        }
+        downloadInfoList.value = downloadInfoBundle
+    }
+
+    fun hide(rowId: Long) {
+        for (i in 0 until downloadInfoBundle.list.size) {
+            val downloadInfo = downloadInfoBundle.list[i]
+            if (rowId == downloadInfo.rowId) {
+                downloadInfoBundle.list.remove(downloadInfo)
+                downloadInfoBundle.notifyType = DownloadInfoBundle.Constants.NOTIFY_ITEM_REMOVED
+                downloadInfoBundle.index = i.toLong()
+                downloadInfoList.value = downloadInfoBundle
+                break
+            }
+        }
+    }
+
+    fun updateRunningItems() {
+        val runningIds = runningDownloadIds
+        if (!runningIds.isEmpty()) {
+            for (i in runningDownloadIds.indices) {
```

Comment:
NIT: it's a bit weird to still use runningDownloadIds after you've declared val runningIds

Reply:
Fixed.

Reply:
remove `runningIds`